ouroboros-consensus-byron/src/Ouroboros/Consensus/Byron/Node/Serialisation.hs

```diff
@@ -77,31 +104,92 @@ instance SerialiseNodeToNode ByronBlock ByronBlock where
+
+  decodeNodeToNode ccfg = \case
+      ByronNodeToNodeVersion1 ->
+        unwrapCBORinCBOR $
+              (flip joinSizeHint fakeByronBlockSizeHint .)
+          <$> decodeUnsizedHeader epochSlots
+      ByronNodeToNodeVersion2 -> decodeDisk ccfg
+
     where
       epochSlots = getByronEpochSlots ccfg
 
 -- | 'Serialised' uses CBOR-in-CBOR by default.
 instance SerialiseNodeToNode ByronBlock (Serialised ByronBlock)
   -- Default instance
 
--- | 'Serialised' uses CBOR-in-CBOR by default.
+-- | Serialise a header streamed from disk.
+--
+-- The bstring in the 'Serialised' matches the output of @WithCodecContext
+-- ByronBlock (Header ByronBlock)@, guaranteed by 'nodeAddHeaderEnvelope'.
+--
+-- * 'ByronNodeToNodeVersion1': we need to maintain backwards compatibility
+--   with the deployed node. We drop the context, giving us the raw header
+--   without the tag indicating whether it's an EBB or not. We need to prepend
+--   that tag (retrieved from the context) again to the raw header. For the
+--   end result, we use CBOR-in-CBOR for backwards compatibility (the old
+--   version used it for the annotations).
+--
+--   May throw 'DropEncodedSizeException' as a pure exception. We have no
+--   other choice, because this function gets used in the codecs whilst
+--   /encoding/ a header, and there we have no natural hook to throw any kind
+--   of decoding errors.
+--
+-- * 'ByronNodeToNodeVersion2': we send exactly the bstring in the
+--   'Serialised' type, which matches exactly what the @SerialiseNodeToNode
+--   ByronBlock (Header ByronBlock)@ instance expects. Note that we /don't/
+--   use CBOR-in-CBOR on the outside. Within the 'WithCodecContext', we
+--   already have CBOR-in-CBOR for the @Serialised (Header ByronBlock)@, we
+--   don't need to wrap the whole thing in CBOR-in-CBOR again.
+--
 instance SerialiseNodeToNode ByronBlock (Serialised (Header ByronBlock)) where
-  encodeNodeToNode _ version = case version of
-      ByronNodeToNodeVersion1 -> encodeWrappedByronHeaderWithoutBlockSize
-      ByronNodeToNodeVersion2 -> encodeWrappedByronHeaderWithBlockSize
-  decodeNodeToNode _ version = case version of
-      ByronNodeToNodeVersion1 -> decodeWrappedByronHeaderWithoutBlockSize
-      ByronNodeToNodeVersion2 -> decodeWrappedByronHeaderWithBlockSize
+  encodeNodeToNode ccfg version = case version of
+      -- Drop the context and add the tag, encode that using CBOR-in-CBOR
+      ByronNodeToNodeVersion1 ->
+            encode
+          . Serialised
+          . addV1Envelope
+          . noErrors
+          . dropNestedTypeEnvelope ccfg
+          . unSerialised
+      ByronNodeToNodeVersion2 -> \(Serialised bs) ->
+        -- Don't use 'encode' on 'Serialised', as that adds CBOR-in-CBOR
+        -- again.
+        CBOR.encodePreEncoded (Lazy.toStrict bs)
+    where
+      noErrors :: Except CBOR.DeserialiseFailure a -> a
+      noErrors = either (throw . DropEncodedSizeException) id . runExcept
```

Comment:
Move exception type to this module.

Reply:
This is now actually nothing to do with Byron anymore. Generalized it to

```haskell
data DropNestedCtxtFailure =
    DropNestedCtxtFailure DeserialiseFailure
  | DropNestedCtxtTrailing Lazy.ByteString
  deriving (Show)

instance Exception DropNestedCtxtFailure
```

in the main lib.